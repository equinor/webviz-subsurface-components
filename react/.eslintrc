{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "extends":  [
    "plugin:react/recommended",
    "eslint:recommended",
    "plugin:prettier/recommended"
  ],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx"],
      "plugins": ["@typescript-eslint"],
      "extends": [
        "plugin:react/recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:prettier/recommended"
      ]
    }
  ],
  "settings": {
    "react": {
      "version": "detect"
    }
  },
  "env": {
    "amd": true,
    "browser": true,
    "es6": true,
    "jasmine": true,
    "node": true
  }
}
