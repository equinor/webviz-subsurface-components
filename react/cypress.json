{
    "baseUrl": "http://localhost:6006/?path=/story/deckglmap--default",
    "video": true
}
