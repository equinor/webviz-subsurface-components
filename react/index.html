<html>
    <head>
        <title>Demo application of webviz-subsurface-components</title>
    </head>
    <body>
        <div id='root'></div>
        <link rel="stylesheet" type="text/css" href="output.css">
        <script src="output.js"></script>
    </body>
</html>
