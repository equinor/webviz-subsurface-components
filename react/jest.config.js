/**
 * The jest tests had an error: SyntaxError: Cannot use import statement outside a module
 * After some investigation, it leads to this page https://github.com/facebook/jest/issues/9395.
 */
module.exports = {
    moduleNameMapper: {
        "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$":
            "<rootDir>/__mocks__/fileMock.js",
        "\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js",
    },
    transform: {
        "\\.(js|ts|jsx|tsx)$": [
            "babel-jest",
            { configFile: "./config/babel.config.json" },
        ],
    },
    transformIgnorePatterns: ["<rootDir>/node_modules/(?!(@webviz)/)"],
};
