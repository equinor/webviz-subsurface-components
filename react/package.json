{
    "name": "@webviz/subsurface-components",
    "description": "Custom Dash components for use in Webviz",
    "repository": {
        "type": "git",
        "url": "git://github.com/equinor/webviz-subsurface-components.git"
    },
    "bugs": {
        "url": "https://github.com/equinor/webviz-subsurface-components/issues"
    },
    "homepage": "https://github.com/equinor/webviz-subsurface-components",
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist/**/*"
    ],
    "scripts": {
        "start": "webpack serve --mode development --devtool inline-source-map --entry ./src/demo/index.js --open",
        "copy-package-json": "copyfiles ./package.json ../webviz_subsurface_components/",
        "transpile": "tsc --project tsconfig.build.json && npm run copy-files",
        "transpile:dash": "tsc --project tsconfig.build.json && rimraf ./dist/components/*/**/",
        "tidy-up": "rimraf ./dist ./webviz_subsurface_components",
        "build:js": "NODE_ENV=production webpack --mode production",
        "build:js-dev": "NODE_ENV=development webpack --mode development",
        "build:js-demo": "NODE_ENV=production webpack --mode production --entry ./src/demo/index.js",
        "build:py": "npm run transpile:dash && copyfiles -a ../webviz_subsurface_components/. ./webviz_subsurface_components/ && copyfiles ./package.json ./webviz_subsurface_components/ && copyfiles -f ./src/lib/assets/VectorDefinitions.json ./webviz_subsurface_components/ && dash-generate-components ./dist/components webviz_subsurface_components -p package.json --ignore '(^index.js|.stories.js)$' && copyfiles -a -f ./webviz_subsurface_components/* ../webviz_subsurface_components/",
        "build": "npm run transpile && npm run build:js && npm run build:js-dev && npm run tidy-up && npm run build:py && npm run tidy-up",
        "typecheck": "tsc --noEmit",
        "format": "eslint --fix *.js *json \"src/**/*.+(ts|tsx|js|jsx|json|css)\"",
        "lint": "eslint *.js *.json \"src/**/*.+(ts|tsx|js|jsx|json|css)\"",
        "validate": "npm run typecheck && npm run lint",
        "setup-deckgl-types": "cd ./node_modules/@danmarshall/deckgl-typings/ && indefinitely-typed --folder deck.gl__aggregation-layers --folder deck.gl__core --folder deck.gl__extensions --folder deck.gl__geo-layers --folder deck.gl__google-maps --folder deck.gl__json --folder deck.gl__layers --folder deck.gl__mapbox --folder deck.gl__mesh-layers --folder deck.gl__react --folder deck.gl --folder luma.gl__constants --folder luma.gl__core --folder luma.gl__gltools --folder luma.gl__webgl-state-tracker --folder luma.gl__webgl --folder math.gl__core --folder math.gl",
        "storybook": "start-storybook -p 6006",
        "build-storybook": "build-storybook",
        "test_perf": "jest _performance",
        "test_correctness": "jest --coverage --testPathIgnorePatterns='_performance'",
        "test": "jest --coverage",
        "test:watch": "npm test -- --watch",
        "build:npm": "rimraf dist && tsc -project tsconfig.build.json --declaration && npm run copy-files",
        "copy-files": "copyfiles package.json dist && copyfiles --up 1 ../README.md dist && copyfiles --exclude **/*.{tsx,jsx,js,ts,snap} --up 2 \"src/lib/**/*\" dist/",
        "prepare": "npm run build:npm",
        "postpack": "npm run tidy-up",
        "prepublishOnly": "npm test && npm run lint",
        "preversion": "npm run lint",
        "test-cypress": "concurrently 'npm run storybook' 'npm run cypress:test' -k -s first",
        "cypress:test": "wait-on http://localhost:6006 && cypress run",
        "cypress:run": "cypress run"
    },
    "author": "Equinor <opensource@equinor.com>",
    "license": "MPL",
    "dependencies": {
        "@emerson-eps/color-tables": "^0.4.22",
        "@equinor/eds-core-react": "^0.12.1",
        "@equinor/eds-icons": "^0.9.1",
        "@equinor/videx-wellog": "^0.7.0",
        "@material-ui/core": "^4.12.3",
        "@material-ui/lab": "^4.0.0-alpha.60",
        "@react-leaflet/core": "^1.1.1",
        "@reduxjs/toolkit": "^1.7.2",
        "@types/d3": "^6.2.0",
        "@types/geojson": "^7946.0.7",
        "@types/react": "^17.0.3",
        "@types/react-dom": "^17.0.3",
        "@types/react-redux": "^7.1.15",
        "@types/react-resize-detector": "^5.0.0",
        "@ungap/structured-clone": "^0.3.4",
        "@vivaxy/png": "^1.3.0",
        "@webviz/core-components": "^0.6.0",
        "ajv": "^7.2.1",
        "clsx": "^1.1.1",
        "convert-units": "^2.3.4",
        "d3": "^5.11.0",
        "d3-format": "^1.4.5",
        "deck.gl": "^8.8.11",
        "deep-equal": "^2.0.5",
        "fast-json-patch": "^3.0.0-1",
        "jsonschema": "^1.4.0",
        "jsverify": "^0.8.4",
        "leaflet": "^1.6.0",
        "leaflet-draw": "^1.0.4",
        "lodash": "^4.17.21",
        "mathjs": "^9.4.2",
        "nebula.gl": "^0.22.3",
        "react": "^17.0.2",
        "react-dom": "^17.0.2",
        "react-dropdown-tree-select": "^2.5.1",
        "react-leaflet": "2.8",
        "react-leaflet-draw": "^0.19.8",
        "react-redux": "^7.2.6",
        "react-resize-detector": "^7.0.0",
        "react-tooltip": "^4.2.21",
        "redux": "^4.0.5",
        "semver": "^7.3.5",
        "styled-components": "^5.3.3",
        "uuid": "^8.3.2"
    },
    "devDependencies": {
        "@actions/core": "^1.9.1",
        "@babel/core": "^7.13.10",
        "@babel/plugin-proposal-class-properties": "^7.13.0",
        "@babel/preset-env": "^7.13.12",
        "@babel/preset-react": "^7.12.13",
        "@babel/preset-typescript": "^7.13.0",
        "@babel/runtime": "^7.13.10",
        "@storybook/addon-docs": "^6.5.9",
        "@storybook/addon-essentials": "^6.5.9",
        "@storybook/addon-links": "^6.5.9",
        "@storybook/addon-storysource": "^6.5.9",
        "@storybook/builder-webpack5": "^6.5.9",
        "@storybook/manager-webpack5": "^6.5.9",
        "@storybook/react": "^6.5.9",
        "@testing-library/cypress": "^8.0.2",
        "@testing-library/jest-dom": "^5.11.9",
        "@testing-library/react": "^12.1.2",
        "@testing-library/user-event": "^13.1.9",
        "@types/cypress-image-snapshot": "^3.1.6",
        "@types/d3-color": "^3.0.2",
        "@types/d3-interpolate": "^3.0.1",
        "@types/jest": "^26.0.20",
        "@types/lodash": "^4.14.171",
        "@types/redux-mock-store": "^1.0.3",
        "@types/semver": "^7.3.8",
        "@types/ungap__structured-clone": "^0.3.0",
        "@types/uuid": "^8.3.0",
        "@typescript-eslint/eslint-plugin": "^4.19.0",
        "@typescript-eslint/parser": "^4.19.0",
        "ace-builds": "^1.4.12",
        "addon-redux": "^2.0.11",
        "babel-jest": "^26.6.3",
        "babel-loader": "^8.1.0",
        "concurrently": "^7.0.0",
        "copyfiles": "^2.0.0",
        "core-js": "^3.9.1",
        "css-loader": "^5.2.0",
        "css-minimizer-webpack-plugin": "^3.0.1",
        "cssnano": "^5.0.5",
        "cypress": "^9.4.1",
        "d3-color": "^3.1.0",
        "dotenv-webpack": "^7.0.2",
        "eslint": "^7.0.0",
        "eslint-config-prettier": "^8.1.0",
        "eslint-plugin-import": "^2.18.2",
        "eslint-plugin-prettier": "^3.1.3",
        "eslint-plugin-react": "^7.19.0",
        "file-loader": "^6.2.0",
        "hex-and-rgba": "^2.0.0",
        "html-webpack-plugin": "^5.3.1",
        "indefinitely-typed": "^1.1.0",
        "jest": "^26.6.3",
        "jest-styled-components": "^7.0.8",
        "json-schema": "^0.4.0",
        "mini-css-extract-plugin": "^1.3.9",
        "node-sass": "^7.0.3",
        "postcss": "^8.3.0",
        "prettier": "^2.5.1",
        "process": "^0.11.10",
        "raw-loader": "^4.0.2",
        "react-ace": "^9.5.0",
        "react-docgen": "^5.3.1",
        "react-test-renderer": "^17.0.2",
        "redux-mock-store": "^1.5.4",
        "sass": "^1.36.0",
        "sass-loader": "^12.1.0",
        "source-map-loader": "^3.0.0",
        "storybook-addon-performance": "^0.16.1",
        "terser-webpack-plugin": "^5.1.1",
        "trim": ">=0.0.3",
        "ts-loader": "^9.2.6",
        "typescript": "4.3.2",
        "url-loader": "^4.1.1",
        "vue-style-loader": "^4.1.3",
        "wait-on": "^6.0.1",
        "webpack": "^5.70.0",
        "webpack-cli": "^4.5.0",
        "webpack-conditional-loader": "^1.0.12",
        "webpack-dev-server": "^4.11.0"
    },
    "engines": {
        "node": ">=14.17",
        "npm": ">=6.14"
    }
}
