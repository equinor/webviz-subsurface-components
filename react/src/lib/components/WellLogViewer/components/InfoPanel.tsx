import React, { Component, ReactNode } from "react";

import { Info } from "./InfoTypes";

interface Props {
    header: string;
    infos: Info[];
    onGroupClick?: (trackId: string | number) => void;
}

function createSeparator() {
    return (
        <tr key={"separator"}>
            {/* Set key prop just for react pleasure. See https://reactjs.org/link/warning-keys for more information */}
            <td colSpan={4}>
                {" "}
                <hr />
            </td>
        </tr>
    );
}

const styleGroupRow = {
    backgroundColor: "#ededed",
    cursor: "pointer",
};

function formatValue(v1: number | string): string {
    if (typeof v1 === "string")
        // discrete log
        return v1;
    if (!Number.isFinite(v1)) return "";
    let v = v1.toPrecision(4);
    if (v.indexOf(".") >= 0) {
        // cut trailing zeroes
        for (;;) {
            let l = v.length;
            if (!l--) break;
            if (v[l] !== "0") break;
            v = v.substring(0, l);
        }
    }
    return v;
}

class InfoPanel extends Component<Props> {
    constructor(props: Props) {
        super(props);
        this.createRow = this.createRow.bind(this);
    }

    onRowClick(
        trackId: string | number /*,
        ev: React.MouseEvent<HTMLTableRowElement>*/
    ): void {
        if (!this.props.onGroupClick) return;
        this.props.onGroupClick(trackId);
    }

    createRow(info: Info): ReactNode {
        if (info.type === "separator")
            // special case
            return createSeparator();

        if (info.groupStart !== undefined) {
            return (
                <tr
                    style={styleGroupRow}
                    key={"_group_" + info.trackId + "." + info.name}
                    onClick={this.onRowClick.bind(this, info.trackId)}
                >
                    <td style={{ color: info.color, fontSize: "small" }}>
                        {
                            info.collapsed
                                ? "\u25BA"
                                : "\u25BC" /*right/down-pointing triangle*/
                        }
                    </td>
                    <td
                        colSpan={3}
                        style={{ fontSize: "small", fontWeight: "bold" }}
                    >
                        {info.name}
                    </td>
                </tr>
            );
        }

        let name = info.name ? info.name : "?";
        if (name.length > 15)
            // compress too long names
            name = name.substring(0, 13) + "â€¦";
        // print long names and values with a smaller font size
        const nameStyle: React.CSSProperties =
            name.length > 10 ? { fontSize: "x-small" } : {};
        const value = formatValue(info.value);
        const valueStyle: React.CSSProperties = {
            width: "90px",
            paddingLeft: "1.5em",
            textAlign: "right",
        };
        if (value.length > 8) valueStyle.fontSize = "x-small";
        return (
            <tr
                key={
                    info.trackId +
                    "." +
                    info.name /*Set unique key prop just for react pleasure*/
                }
            >
                {/*info.type*/}
                <td style={{ color: info.color, fontSize: "small" }}>
                    {"\u2B24" /*big circle*/}
                </td>
                <td style={nameStyle}>{name}</td>
                <td style={valueStyle}>{value}</td>
                <td style={{ paddingLeft: "0.5em" }}>{info.units}</td>
            </tr>
        );
    }

    render(): ReactNode {
        return (
            <div style={{ overflowY: "auto", overflowX: "hidden" }}>
                <fieldset>
                    <legend>{this.props.header}</legend>

                    <table
                        style={{
                            borderSpacing: "0px",
                        }}
                    >
                        <tbody>
                            {this.props.infos.map(this.createRow.bind(this))}
                        </tbody>
                    </table>
                </fieldset>
            </div>
        );
    }
}

export default InfoPanel;
