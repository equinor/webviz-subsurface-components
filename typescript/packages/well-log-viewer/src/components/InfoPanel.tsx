import type { ReactNode } from "react";
import React, { Component } from "react";

import "./rightPanel.scss";

import type { Info } from "./InfoTypes";

interface Props {
    header: string;
    infos: Info[];
    onGroupClick?: (trackId: string | number) => void;
}

function createSeparator(info: Info) {
    return (
        <tr key={"_separator_" + info.trackId + "." + info.name}>
            {/* Set key prop just for react pleasure. See https://reactjs.org/link/warning-keys for more information */}
            <td colSpan={4}>
                {" "}
                <hr />
            </td>
        </tr>
    );
}

function formatValue(value: number): string {
    if (!Number.isFinite(value)) return "";
    if (Number.isInteger(value)) return value.toFixed(0);
    let v = value.toPrecision(4);
    if (v.indexOf(".") >= 0) {
        // cut trailing zeroes
        for (;;) {
            let l = v.length;
            if (!l--) break;
            if (v[l] !== "0") break;
            v = v.substring(0, l);
        }
    }
    return v;
}

const bigCircle = "\u2B24";
const nbsp = "\xA0";
const ellipsis = "\u2026"; //"â€¦";
class InfoPanel extends Component<Props> {
    constructor(props: Props) {
        super(props);
        this.createRow = this.createRow.bind(this);
    }

    onRowClick(
        trackId: string | number /*,
        ev: React.MouseEvent<HTMLTableRowElement>*/
    ): void {
        if (!this.props.onGroupClick) return;
        this.props.onGroupClick(trackId);
    }

    createGroupRow(info: Info): ReactNode {
        return (
            <tr
                className="group-row"
                key={"_group_" + info.trackId + "." + info.name}
                onClick={this.onRowClick.bind(this, info.trackId)}
            >
                <td style={{ color: info.color }}>
                    {
                        info.collapsed
                            ? "\u25BA"
                            : "\u25BC" /*right/down-pointing triangle*/
                    }
                </td>
                <td colSpan={3} className="group-row-name">
                    {info.name}
                </td>
            </tr>
        );
    }

    createRow(info: Info): ReactNode {
        const autoDescreaseFontSize = false;
        if (info.type === "separator")
            // special case
            return createSeparator(info);

        if (info.groupStart !== undefined) return this.createGroupRow(info);

        let name = info.name || "?";
        const tooltip = name;
        // print long names and values with a smaller font size
        const styleInfoName: React.CSSProperties = {};
        let maxLen = 11;
        if (autoDescreaseFontSize && name.length > 10) {
            styleInfoName.fontSize = "x-small";
            maxLen = 16;
        }
        if (name.length > maxLen) {
            // compress too long names
            name = name.substring(0, maxLen - 2) + ellipsis;
        }

        let value = formatValue(info.value);
        if (info.discrete)
            value = info.discrete + (value ? nbsp + "(" + value + ")" : "");
        if (value === "") value = nbsp; // set some text to force an empty line to have the same height as non-empty line
        const styleInfoValue: React.CSSProperties = {};
        if (autoDescreaseFontSize && value.length > 10)
            styleInfoValue.fontSize = "x-small";
        /// style={styleInfoName} title={tooltip!==name? tooltip: undefined}>{name}
        return (
            <tr key={info.trackId + "." + info.name}>
                <td className="row-info" style={{ color: info.color }}>
                    {bigCircle}
                </td>
                <td className="row-name" title={tooltip}></td>
                <td
                    className="row-value"
                    style={styleInfoValue}
                    colSpan={info.discrete ? 2 : 1}
                    title={value}
                ></td>
                {!info.discrete && (
                    <td className="row-units" title={info.units}></td>
                )}
            </tr>
        );
    }

    render(): JSX.Element {
        return (
            <div className="readout">
                <fieldset>
                    <legend>{this.props.header}</legend>

                    <table>
                        <tbody>
                            {this.props.infos?.map(this.createRow.bind(this))}
                        </tbody>
                    </table>
                </fieldset>
            </div>
        );
    }
}

export default InfoPanel;
