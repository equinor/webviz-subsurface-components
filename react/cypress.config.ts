import { defineConfig } from "cypress";
import getCompareSnapshotsPlugin from "cypress-image-diff-js/dist/plugin";

export default defineConfig({
  component: {
    devServer: {
      framework: "create-react-app",
      bundler: "webpack",
    },
    video:false,
    setupNodeEvents(on, config) {
      getCompareSnapshotsPlugin(on, config),
      on('before:browser:launch', (browser = {}, launchOptions) => {
        if (browser.family === 'chromium' && browser.name !== 'electron') {
          launchOptions.args.push('--start-fullscreen')
      
          return launchOptions
        }
        if (browser.name === 'electron') {
          launchOptions.preferences.fullscreen = true
      
          return launchOptions
        }
      })
    }
  },
});
