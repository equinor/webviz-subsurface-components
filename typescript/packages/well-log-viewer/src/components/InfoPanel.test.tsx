import { render } from "@testing-library/react";
import "jest-styled-components";
import "@testing-library/jest-dom";
import React from "react";
import InfoPanel from "./InfoPanel";

describe("Test Info panel", () => {
    it("snapshot test", () => {
        const { container } = render(
            <InfoPanel
                header="Readout"
                infos={[
                    {
                        color: "black",
                        value: 2366,
                        type: "line",
                        trackId: "",
                        name: "DVER",
                        units: "M",
                    },
                    {
                        color: "",
                        value: Number.NaN,
                        type: "separator",
                        trackId: "",
                    },
                    {
                        color: "orange",
                        value: 8.5,
                        type: "line",
                        trackId: "",
                        name: "BDIA",
                        units: "INCH",
                    },
                ]}
            />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
    it("snapshot test when value is infinity", () => {
        const { container } = render(
            <InfoPanel
                header="Readout"
                infos={[
                    {
                        color: "black",
                        value: Infinity,
                        type: "line",
                        trackId: "",
                        name: "DVER",
                        units: "M",
                    },
                ]}
            />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
