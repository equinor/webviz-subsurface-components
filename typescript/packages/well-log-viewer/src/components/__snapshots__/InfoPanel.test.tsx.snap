// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Info panel snapshot test 1`] = `
<div
  class="readout"
>
  <fieldset>
    <legend>
      Readout
    </legend>
    <table>
      <tbody>
        <tr>
          <td
            class="row-info"
            style="color: black;"
          >
            ⬤
          </td>
          <td
            class="row-name"
            title="DVER"
          />
          <td
            class="row-value"
            colspan="1"
            title="2366"
          />
          <td
            class="row-units"
            title="M"
          />
        </tr>
        <tr>
          <td
            colspan="4"
          >
             
            <hr />
          </td>
        </tr>
        <tr>
          <td
            class="row-info"
            style="color: orange;"
          >
            ⬤
          </td>
          <td
            class="row-name"
            title="BDIA"
          />
          <td
            class="row-value"
            colspan="1"
            title="8.5"
          />
          <td
            class="row-units"
            title="INCH"
          />
        </tr>
      </tbody>
    </table>
  </fieldset>
</div>
`;

exports[`Test Info panel snapshot test when value is infinity 1`] = `
<div
  class="readout"
>
  <fieldset>
    <legend>
      Readout
    </legend>
    <table>
      <tbody>
        <tr>
          <td
            class="row-info"
            style="color: black;"
          >
            ⬤
          </td>
          <td
            class="row-name"
            title="DVER"
          />
          <td
            class="row-value"
            colspan="1"
            title=" "
          />
          <td
            class="row-units"
            title="M"
          />
        </tr>
      </tbody>
    </table>
  </fieldset>
</div>
`;
